### PR TITLE
NickAkhmetov/CAT-776 Dataset Relationships diagram

### PR DESCRIPTION
**Pull request by NickAkhmetov**
## Summary

This PR implements the dataset relationships diagram. It is currently displayed under the summary of raw datasets with at least one descendant in provenance that is not a publication or image pyramid. In the future, its position in the page will match the figma designs.

I also included other minor fixes:
- Since the status icons were being defined at runtime, they caused Storybook to flicker.
- The table of contents' section generation code no longer generates sections with the same node key/ID for cases where one dataset has multiple descendants with the same pipeline name. It also correctly handles cases where an invalid descendant has no pipeline name.
- I disabled the Graph tab in the provenance view for datasets that respond to provdata requests with empty objects, as this caused pagewide crashes if clicked on.

### TODOs
- ~~Update summary section of dataset page view~~ - Holding off on this work to avoid interfering with entity header PR after conversation with John
- [x] Integrate the dataset relationships component with the rest of the page
  - Currently appended to the bottom fold of the summary
- [x] Include arrows in diagram edges
- [x] Add storybook story for real prov json example
- [x] Add test coverage for prov -> node conversion logic (if possible)
- [x] Make dataset nodes link to the appropriate hash code
- [x] Update to a more visually pleasing layouting approach (if possible)
- [x] Figure out why the entity nodes sometimes don't appear though the pipeline ones do
    - Appears to have had something to do with dynamically generated styled components - have not been able to reproduce this since fixing that issue.
- Handle various edge cases
	- [x] Publication descendants are not displayed
	- [x] Image pyramid descendants are not displayed
	- [x] Crashes on invalid/error dataset pages
- [x] Fix storybook/jest again 

## Design Documentation/Original Tickets

[Figma](https://www.figma.com/design/GUz7Z37Q3AAMiRyDy7YzWd/Unified-Detail-Pages?node-id=393-57401&m=dev)

[JIRA Subtask](https://hms-dbmi.atlassian.net/browse/CAT-776)

## Testing

- Added Storybook stories
- Tested generated diagrams locally with 6df2f796ad72307d04dc94d688b725c5 (CODEX) and a6a35e2441c6196e74d36c95261df592 (Visium)

## Screenshots/Video
image

image

image

## Checklist

- [x] Code follows the project's coding standards
  - [x] Lint checks pass locally
  - [x] New `CHANGELOG-your-feature-name-here.md` is present in the root directory, describing the change(s) in full sentences.
- [x] Unit tests covering the new feature have been added
- [x] All existing tests pass
- [x] Any relevant documentation in JIRA/Confluence has been updated to reflect the new feature
- [x] Any new functionalities have appropriate analytics functionalities added

## Additional Notes

Please specify any additional information or context relevant to this PR.

### CODE REVIEW

**Comment by NickAkhmetov** [comment]
Got layouting working by adding a consistently sized wrapper to the nodes:
image

File: context/app/static/js/components/detailPage/DatasetRelationships/Legend.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      <Box
+        width={32}
+        height={16}
+        borderRadius={borderRadius}
+        bgcolor={nodeColors[nodeKey]}
+        display="inline-flex"
+        alignItems="center"
+        justifyContent="center"
+      >
+        {Icon && <Icon color="primary" fontSize="0.75rem" width="0.75rem" height="0.75rem" />}
```
Comment (john-conroy) [inline context/app/static/js/components/detailPage/DatasetRelationships/Legend.tsx:28]: What necessitated both `fontSize` and `width` + `height`?

File: context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  const contents = (
+    <Box height="4.125rem" display="flex" alignItems="center">
+      <Stack
+        direction="column"
+        px={2}
+        py={1}
+        borderRadius={rounded ? '1rem' : 0}
+        maxWidth="18rem"
+        bgcolor={bgColor}
+        boxShadow="0px 0px 2px 0px rgba(0, 0, 0, 0.14), 0px 2px 2px 0px rgba(0, 0, 0, 0.12), 0px 1px 3px 0px rgba(0, 0, 0, 0.20)"
```
Comment (john-conroy) [inline context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx:77]: Custom box shadow that wasn't in the theme, I imagine?

File: context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+}: NodeTemplateProps) {
+  // Outer wrapper Box makes sure that nodes are always the same height
+  const contents = (
+    <Box height="4.125rem" display="flex" alignItems="center">
+      <Stack
+        direction="column"
+        px={2}
+        py={1}
+        borderRadius={rounded ? '1rem' : 0}
+        maxWidth="18rem"
```
Comment (john-conroy) [inline context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx:75]: Why these `rem` values?

File: context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+type PipelineNodeProps = Node<PipelineNodeInfo, 'pipeline'>;
+
+function PipelineNode({
+  data: { childDatasets, singleAssay, name, description: _description },
+}: NodeProps<PipelineNodeProps>) {
+  const { pipelineInfo = name, isLoading } = usePipelineInfo(childDatasets);
+  const selectedName = singleAssay ? pipelineInfo : 'Multi Assay Pipeline';
+  const description = singleAssay ? _description : 'Create component datasets';
```
Comment (john-conroy) [inline context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx:113]: Merge into a single ternary?

File: context/app/static/js/components/detailPage/DatasetRelationships/utils.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+  if (currentEntityIsPrimary) {
+    return 'primaryDataset';
+  }
+  if (currentEntityIsComponent) {
+    return 'componentDataset';
+  }
+  return 'processedDataset';
+}
+
+function datasetTypeIsForbidden(datasetType: string) {
```
Comment (john-conroy) [inline context/app/static/js/components/detailPage/DatasetRelationships/utils.ts:60]: Maybe a comment describing what forbidden means in this context?

File: context/app/static/js/components/detailPage/DatasetRelationships/utils.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  if (datasetType === 'Publication') {
+    return true;
+  }
+  if (datasetType.includes('[Image Pyramid]')) {
+    return true;
+  }
+  return false;
+}
+
+function activityTypeIsForbidden(activityType: string) {
```
Comment (john-conroy) [inline context/app/static/js/components/detailPage/DatasetRelationships/utils.ts:70]: Ditto.

File: context/app/static/js/components/detailPage/DatasetRelationships/utils.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+function makeEdge(source: string, target: string): Edge {
+  return {
+    id: `${source}-${target}`,
+    source,
+    target,
+    type: 'straight',
+    markerEnd: {
+      type: MarkerType.ArrowClosed,
+    },
+    style: { color: '#636363' },
```
Comment (john-conroy) [inline context/app/static/js/components/detailPage/DatasetRelationships/utils.ts:86]: I believe this is `secondary.main`?

File: context/app/static/js/hooks/useSoftAssay.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  assaytype: string;
+  'contains-pii': boolean;
+  description: string;
+  'pipeline-shorthand': string;
+  primary: boolean;
+  'vitessce-hints': string[];
+}
+
+export async function fetchSoftAssay({ url, dataset, groupsToken }: SoftAssayRequest) {
+  const response = (await fetch(`${url}/${dataset}`, {
```
Comment (john-conroy) [inline context/app/static/js/hooks/useSoftAssay.tsx:20]: Should we use the `fetcher`?

**Review by john-conroy — APPROVED** [review]
Very clean! A few questions mostly for my benefit, feel free to ignore any.

File: context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx (hunk 3 above)
Comment (NickAkhmetov) [reply context/app/static/js/components/detailPage/DatasetRelationships/nodeTypes.tsx:75]: Mainly to match the measurements from the Figma designs; the outer `Box` is necessary to make the node positions line up (i.e. by making them all the same height without affecting the appearance of the content)

File: context/app/static/js/components/detailPage/DatasetRelationships/Legend.tsx (hunk 1 above)
Comment (NickAkhmetov) [reply context/app/static/js/components/detailPage/DatasetRelationships/Legend.tsx:28]: I believe this was an artifact from troubleshooting the issue with dynamic status icons causing flickering, I'll remove the unnecessary width/height props.

**Pull request merged**